src/vs/workbench/contrib/debug/common/abstractDebugAdapter.ts

```diff
@@ -107,35 +105,72 @@ export abstract class AbstractDebugAdapter implements IDebugAdapter {
 	acceptMessage(message: DebugProtocol.ProtocolMessage): void {
 		if (this.messageCallback) {
 			this.messageCallback(message);
+		} else {
+			this.queue.push(message);
+			if (this.queue.length === 1) {
+				// first item = need to start processing loop
+				this.processQueue();
+			}
 		}
-		else {
-			this.queue.queue(() => {
-				switch (message.type) {
-					case 'event':
-						if (this.eventCallback) {
-							this.eventCallback(<DebugProtocol.Event>message);
-						}
-						break;
-					case 'request':
-						if (this.requestCallback) {
-							this.requestCallback(<DebugProtocol.Request>message);
-						}
-						break;
-					case 'response':
-						const response = <DebugProtocol.Response>message;
-						const clb = this.pendingRequests.get(response.request_seq);
-						if (clb) {
-							this.pendingRequests.delete(response.request_seq);
-							clb(response);
-						}
-						break;
-				}
+	}
+
+	/**
+	 * Returns whether we should insert a timeout between processing messageA
+	 * and messageB. Artificially queueing protocol messages guarantees that any
+	 * microtasks for previous message finish before next message is processed.
+	 * This is essential ordering when using promises anywhere along the call path.
+	 *
+	 * For example, take the following, where `chooseAndSendGreeting` returns
+	 * a person name and then emits a greeting event:
+	 *
+	 * ```
+	 * let person: string;
+	 * adapter.onGreeting(() => console.log('hello', person));
+	 * person = await adapter.chooseAndSendGreeting();
+	 * ```
+	 *
+	 * Because the event is dispatched synchronously, it may fire before person
+	 * is assigned if they're processed in the same task. Inserting a task
+	 * boundary avoids this issue.
+	 */
+	protected needsTaskBoundaryBetween(messageA: DebugProtocol.ProtocolMessage, messageB: DebugProtocol.ProtocolMessage) {
+		return messageA.type !== 'event' || messageB.type !== 'event';
```

Comment:
I think this technically only needs to be looking at one message, like `needsTaskBoundaryBefore(message)`

Reply:
True, good thinking!

Reply:
Actually, I think we need both. We want a boundary between `method return -> event` (e.g. the doc example and test case) but not `event -> event`.

Reply:
Oh, right

Reply:
Oh, right